chains/orchestrator-relays/runtime/dancelight/src/bridge_to_ethereum_config.rs

```diff
@@ -236,6 +242,192 @@ where
+where
+    T: snowbridge_pallet_inbound_queue::Config
+        + pallet_ethereum_token_transfers::Config
+        + pallet_foreign_asset_creator::Config,
+    XcmProcessor: ExecuteXcm<T::RuntimeCall>,
+    XcmWeigher: WeightBounds<T::RuntimeCall>,
+    EthereumLocation: Get<Location>,
+    EthereumNetwork: Get<NetworkId>,
+{
+    fn can_process_message(channel: &Channel, envelope: &Envelope) -> bool {
+        // Ensure that the message is intended for the current channel, para_id and agent_id
+        if let Some(channel_info) =
+            pallet_ethereum_token_transfers::CurrentChannelInfo::<Runtime>::get()
+        {
+            if envelope.channel_id != channel_info.channel_id
+                || channel.para_id != channel_info.para_id
+                || channel.agent_id != channel_info.agent_id
+            {
+                return false;
+            }
+        } else {
+            return false;
+        }
+
+        // Check it is from the right gateway
+        if envelope.gateway != T::GatewayAddress::get() {
+            return false;
+        }
+
+        if let Some(eth_transfer_data) =
+            Self::decode_message_for_eth_transfer(&mut envelope.payload.as_slice())
+        {
+            // Check if the token location is a foreign asset included in ForeignAssetCreator
+            return pallet_foreign_asset_creator::ForeignAssetToAssetId::<Runtime>::get(
+                eth_transfer_data.token_location,
+            )
+            .is_some();
+        }
+
+        return false;
+    }
+
+    fn process_message(_channel: Channel, envelope: Envelope) -> DispatchResult {
+        let eth_transfer_data =
+            Self::decode_message_for_eth_transfer(&mut envelope.payload.as_slice())
+                .ok_or_else(|| DispatchError::Other("unexpected message"))?;
+
+        match eth_transfer_data.destination {
+            Destination::AccountId32 { id: _ } => {
+                Self::process_xcm_local_native_eth_transfer(eth_transfer_data)
+            }
+            // TODO: Add support for container transfers here
+            _ => {
+                return Err(DispatchError::Other(
+                    "container transfers not supported yet",
+                ))
+            }
+        }
+    }
+}
+
+/// Information needed to process an eth transfer message or check its validity.
+pub struct EthTransferData {
+    token_location: Location,
+    destination: Destination,
+    amount: u128,
+}
+
+impl<T, XcmProcessor, XcmWeigher, EthereumLocation, EthereumNetwork>
+    EthTokensLocalProcessor<T, XcmProcessor, XcmWeigher, EthereumLocation, EthereumNetwork>
+where
+    T: frame_system::Config,
+    XcmProcessor: ExecuteXcm<T::RuntimeCall>,
+    XcmWeigher: WeightBounds<T::RuntimeCall>,
+    EthereumLocation: Get<Location>,
+    EthereumNetwork: Get<NetworkId>,
+{
+    /// Retrieve the eth transfer data from the message payload.
+    fn decode_message_for_eth_transfer(payload: &mut &[u8]) -> Option<EthTransferData> {
```

Comment:
Does it work if you do this?

```suggestion
    fn decode_message_for_eth_transfer(mut payload: &[u8]) -> Option<EthTransferData> {
```

Reply:
Yes it does, applied in db34d4d7b71a1ac4d3df49ce526d71f0c61645b6.